### PR TITLE
rgw_sal_motr: [CORTX-30344]PutBucketTagging: Sets tags for a bucket

### PR DESCRIPTION
**Pull request by selvakumaar5496**
Problem:
PutBucketTagging API didn't work with motr as backend. It throws BucketAlreadyExists error while trying to set tags for a bucket

Resolution:
Slight modifications has been done in sal_motr code to enable the API.
Based on exclusive(boolean), it sets flags as M0_OIF_OVERWRITE which resolves BucketAlreadyExists error
```
M0_OIF_OVERWRITE -->For M0_IC_PUT operation, instructs it to silently overwrite existing record with the same key, if any
```

Results:
```
[root@ssc-vm-g2-rhev4-3138 ~]# aws --endpoint=$endpoint s3api put-bucket-tagging --bucket test-buckets --tagging '{"TagSet":[{ "Key": "team", "Value": "c5-solvers" }]}'
[root@ssc-vm-g2-rhev4-3138 ~]# aws --endpoint=$endpoint s3api get-bucket-tagging --bucket test-buckets                                {
    "TagSet": [
        {
            "Key": "team",
            "Value": "c5-solvers"
        }
    ]
}
```

Signed-off-by: Selvakumar <selva.k.nambi@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -726,7 +726,7 @@ int MotrBucket::put_info(const DoutPrefixProvider *dpp, bool exclusive, ceph::re
 
   // Insert bucket instance using bucket's marker (string).
   int rc = store->do_idx_op_by_name(RGW_MOTR_BUCKET_INST_IDX_NAME,
-                                  M0_IC_PUT, tenant_bkt_name, bl, !exclusive);
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:729]: @selvakumaar5496 can you add comment here in review, stating how this change is making the difference

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
Please add problem & resolution sections in PR description (refer https://github.com/Seagate/cortx-rgw/pull/178)

**Review by sachinpunadikar — APPROVED** [review]
+1

**Review by cdeshmukh — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:729]: Hi @cdeshmukh , @sachinpunadikar , @kupranay ,
The above change is incorrect.
```c++
int MotrStore::do_idx_op_by_name(string idx_name, enum m0_idx_opcode opcode,
                                 string key_str, bufferlist &bl, bool update)
```
the last argument of `do_idx_by_name()` function is `update` which is the opposite of `exclusive`.
Meaning `not` of `exclusive` should be passed to `update` arg.

this patch is causing `BucketAlreadyExists` issues in APIs which update the bucket metadata, for example, `put-bucket-quota` or `set-bucket-versioning` etc. 
These APIs are correctly passing `exclusive = false` arg to `MotrBucket::put_info()`, but the wrong interpretation of exclusive arg is causing the error.
```bash
Wed, 4 May 2022 08:50:10 GMT
/admin/bucket
PUT http://ssc-vm-g4-rhev4-0318.colo.seagate.com:8000/admin/bucket?quota&bucket=mybkt&uid=mgwuser&max-size-kb=1024&enabled=true HTTP/1.1
Auth: AWS NB71GXG3O7MMBELODQOO:m4CTpp1L4xQ3IYmdd0p3Vuw+UI0=
HTTP/1.1 409 Conflict

Response Content is: {"Code":"BucketAlreadyExists","HostId":""}
```

Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:729]: For `PutBucketTagging`, the following sal layer implementation is wrong.
```c++
int MotrBucket::merge_and_store_attrs(const DoutPrefixProvider *dpp, Attrs& new_attrs, optional_yield y)
{
  for (auto& it : new_attrs)
    attrs[it.first] = it.second;

  return put_info(dpp, y, ceph::real_time());
}
```
here the `put_info()` should pass `exclusive = false` as an argument.

**Pull request merged**